### DISCUSSION TITLE
Dynamic imports, variables

### DISCUSSION THREAD

**Discussion by ferrettinico** [vercel/next.js#29657] [Help]
While reading about "Dynamic Imports" the documentation states:

[Documentation link](https://nextjs.org/docs/advanced-features/dynamic-import)

_Note: In import('path/to/component'), the path must be explicitly written. It can't be a template string nor a variable. Furthermore the import() has to be inside the dynamic() call for Next.js to be able to match webpack bundles / module ids to the specific dynamic() call and preload them before rendering. dynamic() can't be used inside of React rendering as it needs to be marked in the top level of the module for preloading to work, similar to React.lazy._

But if I do:

`const Nav = dynamic(() => import("../../components/" + (process.env.REPLACE_NAV ? "test" : "all") + "/nav"));`

it actually works! if I leave REPLACE_NAV empty, it will import "/components/all/nav", but if I set REPLACE_NAV it will import "/components/all/nav".

This works even for "next export"

Is the documentation wrong or I am getting something wrong?

Thanks

**Comment by dwbruhn** [comment]
I have the same question. I also tried a template string and it works, so I'm wondering what the "gotcha" is...

**Reply by dwbruhn** [reply]
@ferrettinico Try extracting the path into a variable...

```js
const componentPath = "../../components/" + (process.env.REPLACE_NAV ? "test" : "all") + "/nav";
const Nav = dynamic(() => import(componentPath));
```

Do you get this error in the browser console?

> Critical dependency: the request of a dependency is an expression

That's what I encountered when I extracted the path, and it matches the nuanced behavior described in the webpack docs:
> It is not possible to use a fully dynamic import statement, such as import(foo). Because foo could potentially be any path to any file in your system or project.
> The import() must contain at least some information about where the module is located.

https://webpack.js.org/api/module-methods/#dynamic-expressions-in-import

I'm wondering if the Next.js docs are just trying to play it safe by discouraging people from using template strings at all?

**Reply by dwbruhn** [reply]
@ferrettinico just another update, I discovered that using template strings in dynamic imports was causing an SSR mismatch warning. Going back to fully-specific paths fixed the error.

The Next.js [dynamic import feature](https://github.com/vercel/next.js/blob/canary/packages/next/shared/lib/loadable.js) was built on [react-loadable](https://github.com/jamiebuilds/react-loadable), which itself appears abandoned and doesn't say anything about supporting variables in the import path.

I hope someday Next.js will allow this, since it looks like loadable-components (the alternative to react-loadable) supports it: https://loadable-components.com/docs/dynamic-import/

**Reply by fabien** [reply]
@dwbruhn I was wondering the same thing regarding the "gotchas" in the documentation - this works fine for me: 

https://github.com/vercel/next.js/discussions/39364

**Reply by cecsil** [reply]
Isn't that because environment variables are resolved at build time and the problem is variables that are resolved during runtime,  I mean, webpack or next gets this expression evaluated
